### ISSUE TITLE
Extended goals

### ISSUE THREAD

**Issue by bhack** [naibaf7/libdnn#2]
This repository started targeting convolution and its proximity context.
I'm guessing if we could extend this goal to be a multi stakeholder extended API for dnn.
This require the introduction of other common accellerated kernels/ops in the API (probably starting with porting those that in Opencl Caffe branch). I've opened this to collect feedbacks on the possible roadmap cause I think that the Opencl dnn scenario was historically too fragmented and the standard the facto accellaration API design was roadmapped by cudnn. /cc @edgarriba @hughperkins @gongzg (Intel). I don't /cc amd cause I think its Opencl caffe fork has no more resources allocated and it seems more involved in the  [hccaffe](https://bitbucket.org/multicoreware/hccaffe) effort.

**Comment by edgarriba** [comment]
@bhack thx to take this initiative! I'm in.

The chosen approach for generating the kernels source code makes the library quite singular and attractive to explore more the capabilities by adding additional kernels.

**Comment by naibaf7** [comment]
@bhack Good idea :) I'll be back on improving libDNN after the 14th of July, and will consider which kernels to generalize into libDNN from Caffe.

**Comment by bhack** [comment]
@hughperkins What are you experimenting in https://github.com/hughperkins/gpu-experiments?

**Comment by CNugteren** [comment]
I am also willing to contribute on the kernel development and tuning. In particular if there is something that would be fit as an extension to [my BLAS library](https://github.com/CNugteren/CLBlast) such as a batched GEMM or some other special version of matrix-multiplication.

**Comment by bhack** [comment]
@CNugteren You are welcome.

**Comment by gongzg** [comment]
@bhack @naibaf7 libdnn is a good idea. I will be happy to contribute.

**Comment by bhack** [comment]
@gongzg Nice to have you on board!
@keryell do you have somebody in Xilinx that could be  interested in this?

**Comment by edgarriba** [comment]
welcome everybody!

after this first setup of the group I would like to extend you some ideas and questions that came to my mind during the  design of a proper architecture for [tiny-cnn](https://github.com/edgarriba/tiny-cnn/tree/libdnn) (sorry for the advertisement :smile:) where I want to plug the LibDNN concept and which I think that will be a good starting point to shape this project.
- Should libdnn handle all the OpenCL, CUDA backend stuff or it's the consumer framework who will do it?
- How many kernels and for which platforms, devices we're going to provide support, native CPU, OpenCL CPU/GPU/FPGA/etc, CUDA, CUDNN ?
- How the kernels will be attached, as it is right now or in physical independent sources?

As said this are only thoughts. Feel free to comment them, add more or whatever!

**Comment by naibaf7** [comment]
@edgarriba 
My thoughts:
- Optimal if the framework handles device initialization & memory, but I will provide a simple interface which will initialize the chosen devices automatically. There is no need for new code on this, I will copy the existing initialization from OpenCL Caffe.
- Additionally, I can provide functions to handle device memory (allocation & transfer), but the more optimal path would be to do it in the framework (can reuse memory & be more efficient).
- Initially, the convolution kernel builder will support AMD & nVidia chips, both OpenCL and CUDA. In a second development phase (starting from 15th of July or whenever the SK GT laptop arrives), I will add support for ARM Mali and Intel Skylake, probably with the help from @gongzg.
- I think optimally the kernels can all be generated in the same way that it is now regarding convolutions. The advantage here is that the kernel structure can be modularized, and the autotuner can pick & match parts for optimal performance. However if a kernel is vastly different (FFT, non-GEMM methods?) it will be a separate kernel generator.

**Comment by bhack** [comment]
I'm not convinced with the double approach of the first bullet point.
If we want that "clients" have the responsibility of memory and device handling I don't think that the simplified interface makes sense in libdnn.
I think that maintaining both approach in the API can confuse users.

**Comment by naibaf7** [comment]
@bhack 
I think not, and here is how I plan to avoid confusion: The convolution library interface will stay the same. I will just allow the device-class constructor to be used without previously initialized OpenCL+ViennaCL respectively CUDA context. Furthermore, I will add a method to allocate and copy memory in the device class. That's it, and is not adding confusion I believe.

**Comment by bhack** [comment]
So do you mean  "Optimal if the framework handles device initialization & memory" in term of this simplified interface call?
Probably could we still select an abstracted device list to use client side?

**Comment by naibaf7** [comment]
@bhack 
Well, in case the framework already uses CUDA or OpenCL devices for other kernels/methods/libraries like cuDNN, cuBLAS, clBLAS, CLBlast, it does make much more sense to just use the CUDA or ViennaCL context from an EXISTING context and create a LibDNN-device without initializing a device again...
Just if the framework ONLY needs OpenCL/CUDA for libDNN, then the "simplified" interface will be useful, because it will handle everything.

Thus also the reason why both options are needed.

**Comment by bhack** [comment]
Probably could be a solution to support both use cases. But we need to think how to retrieve the list of devices to let clients select what kind of devices wants to use with the simplified interface.

**Comment by naibaf7** [comment]
@bhack 
Yeah, such a function also already exists in OpenCL Caffe, I'll just offer it in libDNN as well.

**Comment by keryell** [comment]
@bhack Yes of course we have many people at Xilinx interested in DNN libraries in general and this one too...

**Comment by bhack** [comment]
@keryell If you want add someone from Xilinx to this thread you are welcome.

**Comment by hughperkins** [comment]
@bhack re: gpu-experiments.  I am learning :-)  Better late than never I reckon :-D

**Comment by bhack** [comment]
@hughperkins Yes surely late but also AMD (imho one of the big CL player with Intel) quite changed its strategy on Deep Learning. [See a little bit more near to your work](https://bitbucket.org/multicoreware/hctorch)

**Comment by bhack** [comment]
Has anyone benchmarked kernels in [HIP C++ language](https://github.com/GPUOpen-ProfessionalCompute-Tools/HIP)?

**Comment by hughperkins** [comment]
@bhack  So, one of the things I've been learning, is that generally speaking, GPUs tend to obey the laws of physics,  That is, it doesnt matter which language you write things in, the execution time is governed by things like how many warps each sm can hold simultaneously, latency on instructions, and latency on memory fetches, for example.  Things tend to run the same speed in all languages, on the whole.

What do you see as advantages to hcc, in terms of performance?

**Comment by bhack** [comment]
It not was simply in term of performances cause I've not tested it. But there are some interesting things in the [FAQ](https://github.com/GPUOpen-ProfessionalCompute-Tools/HIP/blob/master/docs/markdown/hip_faq.md)

**Comment by hughperkins** [comment]
Yes.  So, basically, as far as I know, if you came across a brand-new library, written in CUDA, hcc should make it dead-easy to port to hcc, since hcc is kind of ... similar :-P  ... to CUDA.  However, it wont magically make things faster or slower, and, specifically, hcc does not in itself contain an implementatoin of cudnn as far as I know?

**Comment by bhack** [comment]
It is not only a porting tool but it was not a proposal.
So on:
Amd is pushing on HCC on caffe and torch forks with third party help.
Nvidia it is stalled on Opencl 1.2 but it is at the head of Opencl/Kronos and cudnn it is still the standard de facto acceleration API. If we could create this alternative with Opencl 1.2 it is OK but I've seen that also porting Neon kernels has created some issue using only Opencl 1.2 features. Or not?

**Comment by hughperkins** [comment]
> I've seen that also porting Neon kernels has created some issue using only Opencl 1.2 features. Or not?

I would say: Not :-).  Basically:
- on NVIDIA, we can use inline assembler, so it's easy to use `ballot`, `popcnt` etc, zero effort
- on all other devices we can use whatever is available on those devices, eg OpenCL 2.0 sub-groups and so on.  Currently, `ballot` is not currently supported in OpenCL 2.0 though.  Looks like AMD GCN ISA does not actually support it? http://amd-dev.wpengine.netdna-cdn.com/wordpress/media/2013/07/AMD_GCN3_Instruction_Set_Architecture.pdf

It's pretty easy to mix and match different OpenCL library versions according to the current hardware targeted. I long ago (well, ok, in last couple of months...) gave up on the idealistic idea that one single kernel could run unmodified everywhere.  Think of it more like writing a webpage that will work across IE, Firefox, etc ...

**Comment by bhack** [comment]
@naibaf7 Can you recap a possible roadmap where you'll we be operative again?

**Comment by naibaf7** [comment]
@bhack 
I am back at work. Currently I have to reinstall my platforms (GTX 1080 just arrived, as well as a platform for Intel Iris (HD 540)).
I will post a roadmap shortly :)

**Comment by bhack** [comment]
@naibaf7 Do you will have also an Intel [Knights Landing](http://www.nextplatform.com/2016/06/21/knights-landing-solid-ground-intels-stake-deep-learning/)?

**Comment by naibaf7** [comment]
@bhack Knights Landing seems to be more in the MKL/CPU space and does not do well (also has limited support) in regards to OpenCL. So no, it is not planned to optimize or work on Knights Landing.

I have enough work upcoming with FP16, Mali/ARM, Intel SKGT 540, Pascal/Hawaii/Polaris as it is :)

**Comment by bhack** [comment]
Probably in Q3 MKL-DNN will be open sourced... we will see.

**Comment by hughperkins** [comment]
> Probably in Q3 MKL-DNN will be open sourced... 

Wow.   Thats ... surprising.  I thought MKL was one of Intels' major pieces of IP?  I remember that one of the advantages of using matlab was that it had MKL as part of it.

**Comment by bhack** [comment]
But I think that with mkl dnn they will open source only the [dnn layers](https://newsroom.intel.com/newsroom/wp-content/uploads/sites/11/2016/06/Intel-ISC16-press-deck-x.pdf).

**Comment by hughperkins** [comment]
Ah, MKL-DNN != MKL.  I see.

**Comment by bhack** [comment]
Every player has its own closed source near metal plan. So we are quite "alone" if we want to go ahead with this :unamused:

**Comment by hughperkins** [comment]
> Every player has its own closed source near metal plan.

I think that's what they need really... you can get quite far in OpenCL.  But, have you seen the level of detail that @scott-gray went into for his winograd kernels?  [this](https://github.com/NervanaSystems/maxas/wiki/SGEMM) describes his gemm implementation, not his winograd implementation, but I think it's fair to say it's likely representative of his winograd implementation.  It looks like registers themselves are arranged in banks, and have bank conflicts, and he spends considerable time using registers in a specific order, in order to avoid such conflicts. https://github.com/NervanaSystems/maxas/wiki/SGEMM#calculating-c-register-banks-and-reuse  Such optimizations are entirely out of reach of OpenCL, at least ,without an optimizing compiler capable of computing them automatically, which is apparently pretty hard to do, otherwise NVIDIA would already have done it.

**Comment by bhack** [comment]
And this are some details of his [Winograd implementation](https://www.nervanasys.com/winograd-2/). Probably we will need machine learning in compilers :grin:

**Comment by hughperkins** [comment]
> And this are some details of his Winograd implementation. 

Ah, thats new information.  Thanks! :-)

> Probably we will need machine learning in compilers :grin:

Yes, maybe :-)  Even if there are complicated underlying mechanisms for certain behaviors, it seems plausible that with sufficient data, a network, or similar, might be able to empirically learn reasonable guidelines for predicting the effect of various layouts, sequencing, and so on.

**Comment by bhack** [comment]
@hughperkins I think that @dividiti probably has something to tell us on this topic. See also http://arxiv.org/abs/1511.02490

**Comment by hughperkins** [comment]
> http://arxiv.org/abs/1511.02490

Interesting.  Although their baseline seems pretty naive?  ie, their baseline is fixed-size workgroups.  I would expect them to use eg Cedric's autotuner as a baseline/comparison for example?  Although they do reference clTune, they dont seem to provide any benchmarks relative to it?

**Comment by bhack** [comment]
For cltune @CNugteren is already here :simile:

**Comment by bhack** [comment]
This is part of a big movement. See http://ctuning.org/cm/wiki/index.php?title=Reproducibility

**Comment by edgarriba** [comment]
@naibaf7 
As mentioned in https://github.com/BVLC/caffe/pull/4421#issuecomment-235021339 by @CNugteren maybe it's a good idea to rely on [https://github.com/CNugteren/CLCudaAPI](https://github.com/CNugteren/CLCudaAPI) and focus libdnn as a pure kernels library in order to take advantage of their advanced work.

Besides, I'd like to propose to embed libdnn in a header to make it more portable. What do you think?

**Comment by naibaf7** [comment]
@edgarriba 
I do not want to port it over to CLCudaAPI at this time, especially since I'm not certain yet in which direction OpenCL Caffe will go at this time, but also because the autotuning, kernel launching and device capability checks are very tightly coupled in there.

As for the header-only: Yes that's an option. What are possible advantages and disadvantages here? ViennaCL and Boost are built as header-only and work well...

**Comment by edgarriba** [comment]
@naibaf7
You know more the low levels details but it's a pity because there's a lot of useful stuff there that probably will be reimplemented (or at least I was thinking to do it). Hope we can balance it with Caffe implementation.

Regarding header-only, for me just having something to directly include with a header it's like a hand in a glove. However, all that glitters is not gold. Check this discussion: http://programmers.stackexchange.com/questions/305618/are-header-only-libraries-more-efficient

**Comment by bhack** [comment]
@naibaf7 We need to know how to go ahead time it is precious on GSoC if we consider libdnn upstream.. Why we cannot start introducing this here? https://github.com/BVLC/caffe/pull/4421#issuecomment-235064468 here

**Comment by bhack** [comment]
If we cannot share a common roadmap on this probably we will adopt this header interface in tinycnn for now. So that we can start to port kernels from caffe this week.

**Comment by naibaf7** [comment]
@bhack 
The discussion there should be sufficient to understand why I will not merge this PR, at least not yet.
CLCudaAPI as you pointed out is another option, but that also needs discussion with the roadmap of C. Nugteren first.

I understand that you need to go ahead faster on GSoC, so please do so and copy the kernels you need from Caffe into your header kernel library, I'm okay with that. The OpenCL kernels are nicely isolated and verified, and you can use whatever API you want (ViennaCL, CLCudaAPI, C++ OpenCL, C OpenCL, ...) to launch them.

**Comment by bhack** [comment]
It is OK.. But probably we cannot gain a critical mass here.

**Comment by naibaf7** [comment]
@bhack 
Yes sorry. I just don't have the time to go forward that fast :)
And I'm not sure how many resources I can gather to help me, i.e. integrating the kernels into libdnn.
The kernels are here: https://github.com/naibaf7/caffe/tree/master/src/caffe/greentea/cl_kernels

**Comment by bhack** [comment]
Yes I know and I understand but please reverse some effort here instead of Caffe if you belive that this could be a really alternative to cudnn. ;)

**Comment by naibaf7** [comment]
@bhack 
If you look at most frameworks that use cuDNN, they actually mainly use the convolutions (and sometimes pooling) and have their own implementations for everything else. Reasons for that were missing parameters, closed-source and faulty implementations. Paired with the fact that from a performance standpoint, it does not help much.
So from my focus, it's worth more to put effort into convolution kernel performance on LibDNN, and the device abstraction / virtual pointers on Caffe first.
Putting more kernels that already exist anyways into LibDNN is nice, ok, but everyone could do it.

**Comment by hughperkins** [comment]
Yes, concur with Fabian, lets keep this clean.  We're not trying to make a new torch/caffe here, we want super magic fast convolutions.

**Comment by bhack** [comment]
Yes but Nvidia needs to invest time on what create most margin. I don't think it cares to let not reinvent the wheel for every layer on that you want to maintain the dataflow on GPU for each new framework we invent ;). I think that many projects will not adopt no mission critical kernels by the closed source nature of cudnn. Nobody has still dimostrated that a good acelleration API, with open source kernels cannot gain momentum with a more comprensive multi stakeholder effort.

**Comment by naibaf7** [comment]
@bhack 
You do not have to reinvent. I linked the kernels above, you can just use that folder from Caffe and you have it all! The boilerplate launch code for these is literally trivial...
@hughperkins will probably agree with this, having implemented not one but two OpenCL DNN frameworks by himself.

**Comment by bhack** [comment]
Nevermind we have a different strategic vision but probably the time is not mature. With every project, with compatible license we could  copy code between project. But not everyone copy code project by project and somebody created libraries, others standardized interfaces like BLAS. 
I don't think that NNPACK has reimplemented Torch or Caffe.
Your vision instead is that this will be something like a very fast gemm alternative.

**Comment by hughperkins** [comment]
mostly, implementation of tanh and so on is not a painpoint.  The biggest challenge facing OpenCL libraries right now, the biggest obstacle to adoption is, if you ask someone, will you use OpenCL or CUDA, they'll say "CUDA libraries are faster".  Thats all down to the convolutional layers, at least, mostly.

**Comment by bhack** [comment]
If you want please close this issue cause was about extended goals.

**Comment by naibaf7** [comment]
@hughperkins Thanks, I agree.
@bhack , see, that's where our visions probably differ. With Intel (GPU team) and hopefully @hughperkins on it, libdnn should be (or become) a collection of fast convolution implementations, and the fastest one per device should be picked automatically.

**Comment by bhack** [comment]
@hughperkins Yes I know it is critical but I want to design API for augmenting ops/kernel coverage. I don't think that somebody stopped to implement Winograd or other tuning techniques here if we expand coverage. But infrastructure it is not ready. So we stay connected here as an evolving optimized upstream convolution effort.

**Comment by bhack** [comment]
In parallel https://github.com/tensorflow/tensorflow/issues/2937#issuecomment-235099718. A quite funny coincidence. ;)

**Comment by bhack** [comment]
It is a mix of closed source and too restricted API design https://github.com/tensorflow/tensorflow/issues/2937#issuecomment-226887605
